### PR TITLE
Classification evaluation

### PR DESCRIPTION
**Pull request by evvahu**
This PR contains an addition to the evaluation script.
- Two methods were included to write per class f1 to a file, and if the (new) flag --confusion_matrix is true then it also writes a confusion_matrix to a csv file.
- the paths were changed to "model_path" +"save_name"

### CODE REVIEW

File: scripts/evaluation.py [hunk 1]
```diff
@@ -44,29 +47,59 @@ def class_performance_ranks(ranker, eval_path):
             f.write(s)
 
 
+def class_performance_classification(path_results, gold_loader, dataloader, eval_path):
```
Comment (Blubberli) [inline scripts/evaluation.py]: dataloader is a missleading parameter name. it is not da dataloader but a dataset.

File: scripts/evaluation.py [hunk 2]
```diff
@@ -44,29 +47,59 @@ def class_performance_ranks(ranker, eval_path):
...
+    preds = np.load(path_results)
+    gold = next(iter(gold_loader))
+    gold = gold["l"].numpy()
+    report = classification_report(gold, preds, output_dict=True)
```
Comment (Blubberli) [inline scripts/evaluation.py]: I think for readability it would be better to not report numbers with like 12 decimals, but I would set it to 3. (parameter digits: 3)

File: scripts/evaluation.py [hunk 3]
```diff
@@ -44,29 +47,59 @@ def class_performance_ranks(ranker, eval_path):
             f.write(s)
 
 
+def class_performance_classification(path_results, gold_loader, dataloader, eval_path):
+    preds = np.load(path_results)
+    gold = next(iter(gold_loader))
+    gold = gold["l"].numpy()
```
Comment (Blubberli) [inline scripts/evaluation.py:65]: if we store these values also in a set we can give it as a parameter to 'classification_report' (target_names). The report will then only include the results for the labels that actually occur in the test data. at the moment I basically have to scroll through the report to find my labels of interest

File: scripts/evaluation.py [hunk 4]
```diff
@@ -44,29 +47,59 @@ def class_performance_ranks(ranker, eval_path):
             f.write(s)
```
Comment (Blubberli) [inline scripts/evaluation.py:52]: method documentation is missing

File: scripts/evaluation.py [hunk 5]
```diff
@@ -84,4 +117,18 @@ def class_performance_ranks(ranker, eval_path):
...
+                conf_path_dev = str(Path(config["model_path"]).joinpath(config["save_name"] + "_confusion_dev.csv"))
+                conf_path_test = str(Path(config["model_path"]).joinpath(config["save_name"] + "_confusion_test.csv"))
+                confusion_matrix_classification(prediction_path_dev, valid_loader, dataset_valid, conf_path_dev)
+                confusion_matrix_classification(prediction_path_test, test_loader, dataset_test, conf_path_test)
+
+        else:
+            class_performance_classification(prediction_path_dev, valid_loader, dataset_valid, eval_path_dev)
+            if argp.confusion_matrix:
+                conf_path_dev = str(Path(config["model_path"]).joinpath(config["save_name"] + "_confusion_dev.csv"))
+                confusion_matrix(prediction_path_dev, valid_loader, dataset_valid, conf_path_dev)
```
Comment (Blubberli) [inline scripts/evaluation.py]: your are trying to call the sklearn method "confusion_matrix". I suppose this is not intended. did you want to call your method "confusion_matrix_classification"?

File: scripts/evaluation.py [hunk 6]
```diff
@@ -44,29 +47,59 @@ def class_performance_ranks(ranker, eval_path):
...
+        if k.isdigit():
+            label = encoder.inverse_transform([int(k)])
+            f.write(str(label) + "\t" + str(v) + "\n")
+        else:
+            f.write(str(k) + "\t" + str(v) + "\n")
+
+    f.close()
+
+
+def confusion_matrix_classification(path_results, gold_loader, dataloader, conf_path):
```
Comment (Blubberli) [inline scripts/evaluation.py]: again, rename parameter "dataloader"

File: scripts/evaluation.py [hunk 7]
```diff
@@ -84,4 +117,18 @@ def class_performance_ranks(ranker, eval_path):
...
-        print("code for classification")
+
+        if config["eval_on_test"]:
+            class_performance_classification(prediction_path_dev, valid_loader, dataset_valid, eval_path_dev)
+            class_performance_classification(prediction_path_test, test_loader, dataset_test, eval_path_test)
+            if argp.confusion_matrix:
+                conf_path_dev = str(Path(config["model_path"]).joinpath(config["save_name"] + "_confusion_dev.csv"))
+                conf_path_test = str(Path(config["model_path"]).joinpath(config["save_name"] + "_confusion_test.csv"))
+                confusion_matrix_classification(prediction_path_dev, valid_loader, dataset_valid, conf_path_dev)
+                confusion_matrix_classification(prediction_path_test, test_loader, dataset_test, conf_path_test)
```
Comment (Blubberli) [inline scripts/evaluation.py]: with more than three arguments, named parameter are more convenient for people who have not written the code.

**Review by Blubberli — CHANGES_REQUESTED** [review]
I have some comments. There is also a bug in the ranking method which I fixed and a method to plot the confusion matrix. I will commit these two changes to this branch.

File: scripts/evaluation.py (hunk 3 above)
Comment (evvahu) [reply scripts/evaluation.py:65]: you mean that we can give it to the confusion matrix? The classification report only inlcudes the labels occurring in the test data.

File: scripts/evaluation.py (hunk 5 above)
Comment (evvahu) [reply scripts/evaluation.py]: seems like I forgot to commit that, I obviously changed that :-)

**Review by Blubberli — APPROVED** [review]
Thanks for the changes. Looks good to me!
